src/lib.rs

```diff
@@ -371,246 +493,346 @@ struct Category<BlockNumber, Moment, AccountId> {
+fn ensure_category_title_is_valid(title: &Vec<u8>) -> Result<(),&'static str> {
+
+    match CATEGORY_TITLE.validate(title.len()) {
+        LengthValidationResult::TooShort => Err(DISPATCHABLE_ERROR_CATEGORY_TITLE_TOO_SHORT),
+        LengthValidationResult::TooLong => Err(DISPATCHABLE_ERROR_CATEGORY_TITLE_TOO_LONG),
+        LengthValidationResult::Success => Ok(())
+    }
+}
+
+fn ensure_category_description_is_valid(description: &Vec<u8>) -> Result<(),&'static str> {
+
+    match CATEGORY_DESCRIPTION.validate(description.len()) {
+        LengthValidationResult::TooShort => Err(DISPATCHABLE_ERROR_CATEGORY_DESCRIPTION_TOO_SHORT),
+        LengthValidationResult::TooLong => Err(DISPATCHABLE_ERROR_CATEGORY_DESCRIPTION_TOO_LONG),
+        LengthValidationResult::Success => Ok(())
+    }
+
+}
+
 impl<T: Trait> Module<T> {
-    // Add public immutables and private mutables.
-    #[allow(dead_code)]
-    fn accumulate_foo(origin: T::Origin, increase_by: BalanceOf<T>) -> dispatch::Result {
-        let _sender = ensure_signed(origin)?;
-
-        let prev = <Foo<T>>::get();
-        // Because Foo has 'default', the type of 'foo' in closure is the raw type instead of an Option<> type.
-        let result = <Foo<T>>::mutate(|foo| {
-            *foo = *foo + increase_by;
-            *foo
-        });
-        assert!(prev + increase_by == result);
+    
+    // Public immutable
+    // - Public interface. These are functions that are `pub` and generally fall into inspector
+    // functions that do not write to storage and operation functions that do.
+    
+    // Private mutables
+    // - Private functions. These are your usual private utilities unavailable to other modules.
+
+    fn current_block_and_time() -> BlockchainTimestamp<T::BlockNumber, T::Moment> {
+
+        BlockchainTimestamp {
+            block: <system::Module<T>>::block_number(),
+            time: <timestamp::Module<T>>::now(),
+        }
+    }
+
+    fn ensure_forum_sudo_set() -> Result<T::AccountId, &'static str> {
+
+        match <ForumSudo<T>>::get() {
+            Some(account_id) => Ok(account_id),
+            None => Err(DISPATCHABLE_ERROR_FORUM_SUDO_NOT_SET)
+        }
+    }
+
+    fn ensure_is_forum_sudo(account_id: &T::AccountId) -> dispatch::Result {
+
+        let forum_sudo_account = Self::ensure_forum_sudo_set()?;
 
+        ensure!(
+            *account_id == forum_sudo_account,
+            DISPATCHABLE_ERROR_ORIGIN_NOT_FORUM_SUDO
+        );
         Ok(())
     }
+
+
+
+/*
+    fn ensure_category_exits(category_id:&CategoryId) -> Result<Category<T::BlockNumber, T::Moment, T::AccountId> , &'static str> {
+
+        if <CategoryById<T>>::exists(category_id) {
+            <CategoryById<T>>::get(category_id)
+        } else {
+            Err("Category does not exist.")
+        }
+
+        Err("dddd")
+    }
+*/
+
+    fn build_category_path(category_id:CategoryId, path: &mut Vec<Category<T::BlockNumber, T::Moment, T::AccountId>>) {
```

Comment:
Consider checking a number of recursive calls. It should be aligned with max category depth.

Reply:
I dont understand what you mean by checking? the termination condition makes sure it ends when it should?

Also recall that, if we make the max depth updatable, e.g. by moderator or governance, then this actually will not even be true, as old category tree structure may not admit to new max value, when max is reduced later.

Reply:
Makes sense!